File: yahtzee.js
import * as modal from "./modal.js";

const StorageKeys = {
  HISTORY: "yahtzee/history",
  QUALIFIED_FOR_TROPHIES: "yahtzee/unlocked trophies",
}

function createNodeFromHtml(html) {
  const template = document.createElement("template");
  template.innerHTML = html.trim();
  return template.content.firstElementChild;
}

function showLoading() {
  document.getElementById("mainframe").classList.add("loading");
}
function finishLoading() {
  const mf = document.getElementById("mainframe");
  requestAnimationFrame(() => {
    mf.style.transform = getComputedStyle(mf).transform;
    mf.classList.remove("loading");
    requestAnimationFrame(() => { mf.style.transform = "rotate3d(0, 0, 0, 0deg)"; });
  });
}
function preloadAssets() {
  for (const src of [
    "one.gif",
    "two.gif",
    "three.gif",
    "four.gif",
    "five.gif",
    "six.gif",
  ]) {
    const img = new Image();
    img.src = src;
  }
}
function prepareYahtzee() {
  const stateful = [
    "selectButton",
    "quitButton",
    "rollButton",
    "resetButton",
  ].map(id => document.getElementById(id));
  stateful.push(...[
    "onesButton",
    "twosButton",
    "threesButton",
    "foursButton",
    "fivesButton",
    "sixesButton",
    "threeOfAKindButton",
    "fourOfAKindButton",
    "fullHouseButton",
    "smallStraightButton",
    "largeStraightButton",
    "chanceButton",
    "yahtzeeButton",
  ].map(id => document.getElementById(id).previousElementSibling));
  for (const elem of stateful) {
    elem.setState = state => {
      if (state === "normal") {
        elem.disabled = false;
        elem.classList.remove("used");
      } else if (state === "disabled") {
        elem.disabled = true;
      } else if (state === "used") {
        elem.disabled = true;
        elem.classList.add("used");
      } else {
        console.warn(`${b.id}.setState called with invalid argument "${state}"`)
        return;
      }
      if (elem.checked) {
        elem.checked = false;
      }
    };
  }
  const images = [
    "die1",
    "die2",
    "die3",
    "die4",
    "die5",
  ].map(id => document.getElementById(id));
  for (const elem of images) {
    elem.changeImage = image => {
      elem.src = image;
      elem.animate(
        [
          {
            backgroundColor: "seashell",
            filter: "saturate(1.5)",
          }, {
            backgroundColor: "",
            filter: "saturate(1)",
          },
        ], 
        {
          duration: 777,
          easing: "ease-in-out",
        }
      );
    }
    elem.setImage = image => {
      elem.src = image;
    }
  }
  const labels = [
    "rollScore",
    "rollCountLabel",
    "subTotalLabel",
    "yahtzeeCountLabel",
    "scoreLabel",
  ].map(id => document.getElementById(id));
  for (const elem of labels) {
    elem.setText = text => elem.textContent = text;
  }
  const checkboxes = [
    "dieHold1",
    "dieHold2",
    "dieHold3",
    "dieHold4",
    "dieHold5",
  ].map(id => document.getElementById(id));
  for (const elem of checkboxes) {
    elem.deselect = () => elem.checked = false;
    elem.addEventListener("change", function() {
      new Audio(`static/sfx/game/${this.checked ? "hold_on" : "hold_off"}.mp3`).play();
    });
  }
  const flashyThings = [
    document.getElementById("subTotalLabel"),
    document.getElementById("yahtzeeButton"),
  ];
  for (const elem of flashyThings) {
    elem.flash = () => elem.animate(
      [
        { outline: "none" },
        { outline: "2px solid red" },
        { outline: "2px solid red" },
      ], 
      {
        duration: 120,
        iterations: 10,
        easing: "steps(2)",
      }
    );
  }
  document.getElementById("subTotalLabel").oldFlash = document.getElementById("subTotalLabel").flash;
  document.getElementById("subTotalLabel").flash = function() {
    this.oldFlash();
    navigator.vibrate?.(100);
    new Audio("static/sfx/game/subtotal_bonus.mp3").play();
  };
  document.getElementById("yahtzeeButton").celebrate = () => {
    navigator.vibrate?.([200, 50, 100]);
    new Audio("static/sfx/game/success.mp3").play();
  };
  document.getElementById("yahtzeeButton").fail = () => {
    new Audio("static/sfx/game/failure.mp3").play();
  };
  document.getElementById("subTotalLabel").setEmphasis = function(on) { this.style.fontStyle = on ? "italic" : "" };
  function gameOver(score, yahtzeeCount, gotBonus) {
    modal.gameOver(
      document.body,
      () => document.getElementById("resetButton").click(), 
      {
        score,
        yahtzeeCount,
        gotBonus,
      }
    );
    saveHistory(score, yahtzeeCount, gotBonus);
    const qualified = score >= 100 || yahtzeeCount > 0;
    if (qualified && !document.getElementById("trophies")) {
      localStorage.setItem(StorageKeys.QUALIFIED_FOR_TROPHIES, "true story");
      createTrophiesButton();
    }
  }
  function saveHistory(score, yahtzeeCount, gotBonus) {
    const saved = JSON.parse(localStorage.getItem(StorageKeys.HISTORY));
    history = Array.isArray(saved) ? saved : [];
    history.push({
      t: Date.now(),
      p: score,
      y: yahtzeeCount,
      s: gotBonus,
    });
    localStorage.setItem(StorageKeys.HISTORY, JSON.stringify(history));
  }
  function bonusYahtzee(on) {
    const target = document.getElementById("scoreLabel");
    if (document.getElementById("bonusYahtzee")) {
      document.getElementById("bonusYahtzee").remove();
    }
    if (on) {
      target.before(createNodeFromHtml(`
        <div id="bonusYahtzee">
          <div style="font-style: italic;">Bonus Yahtzee!</div>
          <div>Joker Rules apply</div>
        </div>
      `));
    }
  }
  function damageNumber(damage) {
    const target = document.getElementById("scoreLabel");
    const dn = createNodeFromHtml(`
      <div class="damageNumber">
        <span>${damage}</span>
      </div>
    `);
    target.before(dn);
    setTimeout(() => {
      dn.remove();
    }, 2000);
  }
  const yobject = {
    photoList: ["one.gif", "two.gif", "three.gif", "four.gif", "five.gif", "six.gif", "sel1.gif", "sel2.gif", "sel3.gif", "sel4.gif", "sel5.gif", "sel6.gif", "sel7.gif", "sel8.gif", "sel9.gif", "sel10.gif", "sel11.gif", "sel12.gif", "sel13.gif", "yahtzee.gif", "cards.gif"],
    yPhotoList: [
      "one_y.gif",
      "two_y.gif",
      "three_y.gif",
      "four_y.gif",
      "five_y.gif",
      "six_y.gif",
    ],
    onesButton: document.getElementById("onesButton").previousElementSibling,
    twosButton: document.getElementById("twosButton").previousElementSibling,
    threesButton: document.getElementById("threesButton").previousElementSibling,
    foursButton: document.getElementById("foursButton").previousElementSibling,
    fivesButton: document.getElementById("fivesButton").previousElementSibling,
    sixesButton: document.getElementById("sixesButton").previousElementSibling,
    threeOfAKindButton: document.getElementById("threeOfAKindButton").previousElementSibling,
    fourOfAKindButton: document.getElementById("fourOfAKindButton").previousElementSibling,
    fullHouseButton: document.getElementById("fullHouseButton").previousElementSibling,
    smallStraightButton: document.getElementById("smallStraightButton").previousElementSibling,
    largeStraightButton: document.getElementById("largeStraightButton").previousElementSibling,
    chanceButton: document.getElementById("chanceButton").previousElementSibling,
    yahtzeeButton: document.getElementById("yahtzeeButton").previousElementSibling,
    yahtzeeButton2: document.getElementById("yahtzeeButton"),
    selectButton: document.getElementById("selectButton"),
    quitButton: document.getElementById("quitButton"),
    rollButton: document.getElementById("rollButton"),
    resetButton: document.getElementById("resetButton"),
    die1: document.getElementById("die1"),
    die2: document.getElementById("die2"),
    die3: document.getElementById("die3"),
    die4: document.getElementById("die4"),
    die5: document.getElementById("die5"),
    rollScore: document.getElementById("rollScore"),
    rollCountLabel: document.getElementById("rollCountLabel"),
    subTotalLabel: document.getElementById("subTotalLabel"),
    yahtzeeCountLabel: document.getElementById("yahtzeeCountLabel"),
    scoreLabel: document.getElementById("scoreLabel"),
    dieHold1: document.getElementById("dieHold1"),
    dieHold2: document.getElementById("dieHold2"),
    dieHold3: document.getElementById("dieHold3"),
    dieHold4: document.getElementById("dieHold4"),
    dieHold5: document.getElementById("dieHold5"),
    gameOver,
    bonusYahtzee,
    showIncrement: damageNumber,
  };
  return yobject;
}
async function runYahtzee() {
  const oldTitle = document.title;
  document.title = "Loading…";
  pyodide = await loadPyodide();
  document.title = oldTitle;
  await pyodide.runPython(await (await fetch("Yahtzee.py")).text());
  attachGameHandlers();
}
function attachGameHandlers() {
  document.getElementById("quitButton").addEventListener("click", () => {
    window.location = "/";
    new Audio("static/sfx/game/click.mp3").play();
  });
  document.getElementById("rollButton").addEventListener("click", () => {
    pyodide.globals.get("roll")();
    new Audio("static/sfx/game/click.mp3").play();
  });
  document.getElementById("resetButton").addEventListener("click", () => {
    pyodide.globals.get("resetGame")();
    new Audio("static/sfx/game/click.mp3").play();
  });
  document.getElementById("selectButton").addEventListener("click", () => {
    pyodide.globals.get("select")();
    new Audio("static/sfx/game/plop.mp3").play();
  });
  
  [
    "onesButton",
    "twosButton",
    "threesButton",
    "foursButton",
    "fivesButton",
    "sixesButton",
    "threeOfAKindButton",
    "fourOfAKindButton",
    "fullHouseButton",
    "smallStraightButton",
    "largeStraightButton",
    "chanceButton",
    "yahtzeeButton",
  ].forEach(id => document.getElementById(id).previousElementSibling.addEventListener("click", () => {
    pyodide.globals.get(id.replace("Button", ""))();
    new Audio("static/sfx/game/select.mp3").play();
  }));
}
function attachMetaHandlers() {
  attachZoomHandler();
  attachHelpHandler();
  attachDesktopHandler();
  attachMobileHandler();
}
function attachZoomHandler() {
  const target = document.querySelector("main");
  const targetValue = "0.8";
  document.querySelector(".zoom").addEventListener("click", function() {
    this.classList.add("stay");
    target.style.transformOrigin = "top";
    target.style.scale = target.style.scale == targetValue ? "" : targetValue;
    new Audio("static/sfx/game/plop.mp3").play();
  });
}
function attachHelpHandler() {
  document.querySelector(".help").addEventListener("click", () => {
    modal.help(document.body);
    new Audio("static/sfx/game/bubbles.mp3").play();
  });
}
const VIEW_MODE_PARAM = "view";
const ViewModes = Object.freeze({
  DESKTOP: "desktop",
  MOBILE: "mobile",
});
function attachDesktopHandler() {
  document.querySelector(".desktop")?.addEventListener("click", () => {
    new Audio("static/sfx/game/click.mp3").play();
    if (!confirm("Reload Maple Yahtzee in Desktop Mode?")) return;
    const url = new URL("./", window.location.href);
    url.searchParams.set(VIEW_MODE_PARAM, ViewModes.DESKTOP);
    window.location.href = url.href;
  });
}
function attachMobileHandler() {
  document.querySelector(".mobile")?.addEventListener("click", () => {
    new Audio("static/sfx/game/click.mp3").play();
    if (!confirm("Reload Maple Yahtzee in Mobile Mode?")) return;
    const url = new URL("./", window.location.href);
    url.searchParams.set(VIEW_MODE_PARAM, ViewModes.MOBILE);
    window.location.href = url.href;
  });
}
async function adjustMode() {
  const viewMode = new URLSearchParams(window.location.search).get(VIEW_MODE_PARAM);
  const isMobile = navigator.userAgentData?.mobile || /Mobi|Android|iPhone|iPad|iPod/i.test(navigator.userAgent);
  if (viewMode === ViewModes.MOBILE || viewMode === null && isMobile) {
    const mobileContent = new DOMParser().parseFromString(await (await fetch("mobile.html")).text(), "text/html");
    document.head.appendChild(Object.assign(document.createElement("meta"), {
      name: "viewport",
      content: "width=device-width, initial-scale=1.0",
    }));
    document.head.appendChild(Object.assign(document.createElement("link"), {
      rel: "stylesheet",
      type: "text/css",
      href: "mobile.css",
    }));
    document.querySelector("main").replaceWith(mobileContent.querySelector("main"));
  } else if (viewMode === ViewModes.DESKTOP) {
    document.querySelector(".auxiliary").prepend(createNodeFromHtml(`
      <button class="mobile" title="Mobile Mode">📱</button>
    `));
  }
}
function createTrophiesButton() {
  if (!localStorage.getItem(StorageKeys.QUALIFIED_FOR_TROPHIES)) {
    return;
  }
  const firstSibling = document.getElementById("mainframe");
  firstSibling.parentNode.append(createNodeFromHtml(`
    <div class="trophiesContainer">
      <button id="trophies">Trophies</button>
    </div>
  `));
  attachTrophiesHandler();
}
function attachTrophiesHandler() {
  document.getElementById("trophies").addEventListener("click", () => {
    modal.trophies(document.body, history);
    new Audio("static/sfx/game/bubbles.mp3").play();
  });
}
try {
  await adjustMode();
} catch (error) {
  console.warn(`Failed to load view mode!`, error);
}
attachMetaHandlers();
let history = JSON.parse(localStorage.getItem(StorageKeys.HISTORY));
createTrophiesButton();
preloadAssets();
globalThis.yobject = prepareYahtzee();
var pyodide;
showLoading();
runYahtzee().then(finishLoading);

File: Yahtzee.py
from js import yobject
import random

def gameOver():
    yobject.rollButton.setState("disabled")
    yobject.gameOver(totalScore, yahtzeeCount, subTotalBonus)

def diceReset():
    global rollCount
    global potentialScore
    global die1value
    global die2value
    global die3value
    global die4value
    global die5value
    potentialScore = 0
    rollCount = 0
    yobject.die1.setImage(yobject.photoList[20])
    yobject.die2.setImage(yobject.photoList[20])
    yobject.die3.setImage(yobject.photoList[20])
    yobject.die4.setImage(yobject.photoList[20])
    yobject.die5.setImage(yobject.photoList[20])
    die1value=0
    die2value=0
    die3value=0
    die4value=0
    die5value=0
    yobject.dieHold1.deselect()
    yobject.dieHold2.deselect()
    yobject.dieHold3.deselect()
    yobject.dieHold4.deselect()
    yobject.dieHold5.deselect()
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    yobject.rollButton.setState("normal")
    yobject.rollCountLabel.setText("Rolls Made: "+str(rollCount))
    selectionDisable()

def select():
    global totalScore
    global selectionCount
    global subTotal
    global yahtzeeCount
    global subTotalBonus
    global usedYahtzee
    if rollUsed == 0:
        return
    if rollUsed == 1:
        yobject.onesButton.setState("used")
        subTotal += potentialScore
    elif rollUsed == 2:
        yobject.twosButton.setState("used")
        subTotal += potentialScore
    elif rollUsed == 3:
        yobject.threesButton.setState("used")
        subTotal += potentialScore
    elif rollUsed == 4:
        yobject.foursButton.setState("used")
        subTotal += potentialScore
    elif rollUsed == 5:
        yobject.fivesButton.setState("used")
        subTotal += potentialScore
    elif rollUsed == 6:
        yobject.sixesButton.setState("used")
        subTotal += potentialScore
    elif rollUsed == 7:
        yobject.threeOfAKindButton.setState("used")
    elif rollUsed == 8:
        yobject.fourOfAKindButton.setState("used")
    elif rollUsed == 9:
        yobject.fullHouseButton.setState("used")
    elif rollUsed == 10:
        yobject.smallStraightButton.setState("used")
    elif rollUsed == 11:
        yobject.largeStraightButton.setState("used")
    elif rollUsed == 12:
        yobject.chanceButton.setState("used")
    elif rollUsed == 13:
        yobject.yahtzeeButton.setState("used")
        usedYahtzee = True
    selectionCount += 1
    totalScore += potentialScore
    if bonusYahtzee:
        totalScore += 100
        yahtzeeCount += 1
    if subTotal >= 63 and subTotalBonus == False:
        totalScore += 35
        subTotalBonus = True
        yobject.subTotalLabel.flash();
        yobject.subTotalLabel.setEmphasis(True);
    yobject.yahtzeeCountLabel.setText("Yahtzees: "+str(yahtzeeCount))
    yobject.subTotalLabel.setText("Sub Total: "+str(subTotal))
    yobject.scoreLabel.setText("SCORE "+str(totalScore))
    
    usedSelections.append(rollUsed)
    diceReset()
    yobject.bonusYahtzee(False)
    if selectionCount == 13:
        gameOver()


def ones():
    global rollUsed
    global potentialScore
    potentialScore = 0
    if die1value == 1:
        potentialScore += 1
    if die2value == 1:
        potentialScore += 1
    if die3value == 1:
        potentialScore += 1
    if die4value == 1:
        potentialScore += 1
    if die5value == 1:
        potentialScore += 1
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 1
    yobject.selectButton.setState("normal")

def twos():
    global rollUsed
    global potentialScore
    potentialScore = 0
    if die1value == 2:
        potentialScore += 2
    if die2value == 2:
        potentialScore += 2
    if die3value == 2:
        potentialScore += 2
    if die4value == 2:
        potentialScore += 2
    if die5value == 2:
        potentialScore += 2
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 2
    yobject.selectButton.setState("normal")

def threes():
    global rollUsed
    global potentialScore
    potentialScore = 0
    if die1value == 3:
        potentialScore += 3
    if die2value == 3:
        potentialScore += 3
    if die3value == 3:
        potentialScore += 3
    if die4value == 3:
        potentialScore += 3
    if die5value == 3:
        potentialScore += 3
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 3
    yobject.selectButton.setState("normal")

def fours():
    global rollUsed
    global potentialScore
    potentialScore = 0
    if die1value == 4:
        potentialScore += 4
    if die2value == 4:
        potentialScore += 4
    if die3value == 4:
        potentialScore += 4
    if die4value == 4:
        potentialScore += 4
    if die5value == 4:
        potentialScore += 4
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 4
    yobject.selectButton.setState("normal")

def fives():
    global rollUsed
    global potentialScore
    potentialScore = 0
    if die1value == 5:
        potentialScore += 5
    if die2value == 5:
        potentialScore += 5
    if die3value == 5:
        potentialScore += 5
    if die4value == 5:
        potentialScore += 5
    if die5value == 5:
        potentialScore += 5
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 5
    yobject.selectButton.setState("normal")

def sixes():
    global rollUsed
    global potentialScore
    potentialScore = 0
    if die1value == 6:
        potentialScore += 6
    if die2value == 6:
        potentialScore += 6
    if die3value == 6:
        potentialScore += 6
    if die4value == 6:
        potentialScore += 6
    if die5value == 6:
        potentialScore += 6
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 6
    yobject.selectButton.setState("normal")

def threeOfAKind():
    global rollUsed
    global potentialScore
    potentialScore = 0
    if (die1value == die2value == die3value)\
or (die1value == die2value == die4value)\
or (die1value == die2value == die5value)\
or (die1value == die3value == die4value)\
or (die1value == die3value == die5value)\
or (die1value == die4value == die5value)\
or (die2value == die3value == die4value)\
or (die2value == die3value == die5value)\
or (die2value == die4value == die5value)\
or (die3value == die4value == die5value)\
or bonusYahtzee:
        potentialScore = die1value + die2value + die3value + die4value + die5value
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 7
    yobject.selectButton.setState("normal")

def fourOfAKind():
    global rollUsed
    global potentialScore
    potentialScore = 0
    if (die1value == die2value == die3value == die4value)\
or (die1value == die2value == die3value == die5value)\
or (die1value == die2value == die4value == die5value)\
or (die1value == die3value == die4value == die5value)\
or (die2value == die3value == die4value == die5value)\
or bonusYahtzee:
        potentialScore = die1value + die2value + die3value + die4value + die5value
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 8
    yobject.selectButton.setState("normal")

def fullHouse():
    global rollUsed
    global potentialScore
    potentialScore = 0
    if (((die1value == die2value) and (die3value == die4value == die5value))\
or ((die1value == die3value) and (die2value == die4value == die5value))\
or ((die1value == die4value) and (die2value == die3value == die5value))\
or ((die1value == die5value) and (die2value == die4value == die3value))\
or ((die2value == die3value) and (die1value == die4value == die5value))\
or ((die2value == die4value) and (die1value == die3value == die5value))\
or ((die2value == die5value) and (die1value == die4value == die3value))\
or ((die3value == die4value) and (die1value == die2value == die5value))\
or ((die3value == die5value) and (die1value == die2value == die4value))\
or ((die4value == die5value) and (die1value == die2value == die3value)))\
and ((die1value == die2value == die3value == die4value == die5value) == False)\
or bonusYahtzee:
        potentialScore = 25
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 9
    yobject.selectButton.setState("normal")

def smallStraight():
    global rollUsed
    global potentialScore
    potentialScore = 0
    if (((die1value == 1) or (die2value == 1) or (die3value == 1) or (die4value == 1) or (die5value == 1))\
and ((die1value == 2) or (die2value == 2) or (die3value == 2) or (die4value == 2) or (die5value == 2))\
and ((die1value == 3) or (die2value == 3) or (die3value == 3) or (die4value == 3) or (die5value == 3))\
and ((die1value == 4) or (die2value == 4) or (die3value == 4) or (die4value == 4) or (die5value == 4)))\
or (((die1value == 2) or (die2value == 2) or (die3value == 2) or (die4value == 2) or (die5value == 2))\
and ((die1value == 3) or (die2value == 3) or (die3value == 3) or (die4value == 3) or (die5value == 3))\
and ((die1value == 4) or (die2value == 4) or (die3value == 4) or (die4value == 4) or (die5value == 4))\
and ((die1value == 5) or (die2value == 5) or (die3value == 5) or (die4value == 5) or (die5value == 5)))\
or (((die1value == 3) or (die2value == 3) or (die3value == 3) or (die4value == 3) or (die5value == 3))\
and ((die1value == 4) or (die2value == 4) or (die3value == 4) or (die4value == 4) or (die5value == 4))\
and ((die1value == 5) or (die2value == 5) or (die3value == 5) or (die4value == 5) or (die5value == 5))\
and ((die1value == 6) or (die2value == 6) or (die3value == 6) or (die4value == 6) or (die5value == 6)))\
or bonusYahtzee:
        potentialScore = 30
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 10
    yobject.selectButton.setState("normal")

def largeStraight():
    global rollUsed
    global potentialScore
    potentialScore = 0
    if (((die1value == 1) or (die2value == 1) or (die3value == 1) or (die4value == 1) or (die5value == 1))\
and ((die1value == 2) or (die2value == 2) or (die3value == 2) or (die4value == 2) or (die5value == 2))\
and ((die1value == 3) or (die2value == 3) or (die3value == 3) or (die4value == 3) or (die5value == 3))\
and ((die1value == 4) or (die2value == 4) or (die3value == 4) or (die4value == 4) or (die5value == 4))\
and ((die1value == 5) or (die2value == 5) or (die3value == 5) or (die4value == 5) or (die5value == 5)))\
or (((die1value == 2) or (die2value == 2) or (die3value == 2) or (die4value == 2) or (die5value == 2))\
and ((die1value == 3) or (die2value == 3) or (die3value == 3) or (die4value == 3) or (die5value == 3))\
and ((die1value == 4) or (die2value == 4) or (die3value == 4) or (die4value == 4) or (die5value == 4))\
and ((die1value == 5) or (die2value == 5) or (die3value == 5) or (die4value == 5) or (die5value == 5))\
and ((die1value == 6) or (die2value == 6) or (die3value == 6) or (die4value == 6) or (die5value == 6)))\
or bonusYahtzee:
        potentialScore = 40
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 11
    yobject.selectButton.setState("normal")

def chance():
    global rollUsed
    global potentialScore
    potentialScore = die1value + die2value + die3value + die4value + die5value
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 12
    yobject.selectButton.setState("normal")

def yahtzee():
    global rollUsed
    global potentialScore
    global yahtzeeCount
    potentialScore = 0
    if die1value == die2value == die3value == die4value == die5value and die1value != 0:
        potentialScore = 50
        yahtzeeCount += 1
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    rollUsed = 13
    yobject.selectButton.setState("normal")

def roll():
    global rollCount
    global rollUsed
    global bonusYahtzee
    global die1value
    global die2value
    global die3value
    global die4value
    global die5value
    potentialScore = 0
    rollUsed = 0
    bonusYahtzee = False
    yobject.bonusYahtzee(False)
    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    if yobject.dieHold1.checked == False or die1value == 0:
        number = random.randrange(1,7)
        yobject.die1.setImage(yobject.photoList[number-1])
        die1value=number
    if yobject.dieHold2.checked == False or die2value == 0:
        number = random.randrange(1,7)
        yobject.die2.setImage(yobject.photoList[number-1])
        die2value=number
    if yobject.dieHold3.checked == False or die3value == 0:
        number = random.randrange(1,7)
        yobject.die3.setImage(yobject.photoList[number-1])
        die3value=number
    if yobject.dieHold4.checked == False or die4value == 0:
        number = random.randrange(1,7)
        yobject.die4.setImage(yobject.photoList[number-1])
        die4value=number
    if yobject.dieHold5.checked == False or die5value == 0:
        number = random.randrange(1,7)
        yobject.die5.setImage(yobject.photoList[number-1])
        die5value=number
    if die1value == die2value == die3value == die4value == die5value:
        if yahtzeeCount == 0:
            if not usedYahtzee:
                yobject.yahtzeeButton.flash()
        else:
            bonusYahtzee = True
            yobject.bonusYahtzee(True)
    rollCount += 1
    yobject.rollCountLabel.setText("Rolls Made: "+str(rollCount))
    rollDisable()
    showAvailableSelections()

def rollDisable():
    if rollCount == 3:
        yobject.rollButton.setState("disabled")

def selectionDisable():
    yobject.onesButton.setState("disabled")
    yobject.twosButton.setState("disabled")
    yobject.threesButton.setState("disabled")
    yobject.foursButton.setState("disabled")
    yobject.fivesButton.setState("disabled")
    yobject.sixesButton.setState("disabled")
    yobject.threeOfAKindButton.setState("disabled")
    yobject.fourOfAKindButton.setState("disabled")
    yobject.fullHouseButton.setState("disabled")
    yobject.smallStraightButton.setState("disabled")
    yobject.largeStraightButton.setState("disabled")
    yobject.chanceButton.setState("disabled")
    yobject.yahtzeeButton.setState("disabled")
    yobject.selectButton.setState("disabled")

def showAvailableSelections():
    yobject.selectButton.setState("disabled")
    buttons = [
        yobject.onesButton,
        yobject.twosButton,
        yobject.threesButton,
        yobject.foursButton,
        yobject.fivesButton,
        yobject.sixesButton,
        yobject.threeOfAKindButton,
        yobject.fourOfAKindButton,
        yobject.fullHouseButton,
        yobject.smallStraightButton,
        yobject.largeStraightButton,
        yobject.chanceButton,
        yobject.yahtzeeButton,
    ]
    for selection in range(1, 13 + 1):
        buttons[selection - 1].setState("disabled" if selection in usedSelections else "normal")

    if not bonusYahtzee:
        return
    if die1value not in usedSelections:
        for button in buttons:
            if button != buttons[die1value - 1]:
                button.setState("disabled")
    elif any(selection not in usedSelections for selection in range(7, 14)):
        for i in range(0, 6):
            buttons[i].setState("disabled")
    else:
        for i in range(6, 13):
            buttons[i].setState("disabled")

def resetGame():
    global rollCount
    global potentialScore
    global totalScore
    global selectionCount
    global yahtzeeCount
    global subTotal
    global bonusYahtzee
    global subTotalBonus
    global usedYahtzee
    rollCount=0
    potentialScore=0
    totalScore=0
    selectionCount=0
    subTotal=0
    bonusYahtzee=False
    yobject.bonusYahtzee(False)
    subTotalBonus=False
    yahtzeeCount=0
    usedYahtzee=False
    usedSelections.clear()

    yobject.rollScore.setText("Roll Value: "+str(potentialScore))
    yobject.scoreLabel.setText("SCORE "+str(totalScore))
    yobject.yahtzeeCountLabel.setText("Yahtzees: "+str(yahtzeeCount))
    yobject.subTotalLabel.setText("Sub Total: "+str(subTotal))
    yobject.subTotalLabel.setEmphasis(False);
    
    diceReset()

#main
rollCount = 0
potentialScore = 0
totalScore = 0
selectionCount = 0
subTotal = 0
yahtzeeCount = 0
bonusYahtzee = False
subTotalBonus = False

#new
die1value = 0
die2value = 0
die3value = 0
die4value = 0
die5value = 0
usedYahtzee = False
usedSelections = []

diceReset()
